File: examples/counter/iOS/CounterApp/http.swift
import SharedTypes
import SwiftUI

enum HttpError : Error {
    case generic(Error)
    case message(String)
}

func httpRequest(_ request: HttpRequest) async -> Result<HttpResponse, HttpError> {
    var req = URLRequest(url: URL(string: request.url)!)
    req.httpMethod = request.method
    
    for header in request.headers {
        req.addValue(header.value, forHTTPHeaderField: header.name)
    }
    
    do {
        let (data, response) = try await URLSession.shared.data(for: req)
        if let httpResponse = response as? HTTPURLResponse {
            let status = UInt16(httpResponse.statusCode)
            let body = [UInt8](data)
            return .success(HttpResponse(status: status, headers: [], body: body))
        } else {
            return .failure(.message("bad response"))
        }
    }
    catch {
        return .failure(.generic(error))
    }
}

File: examples/counter/web-remix/app/http.ts
import type { HttpRequest, HttpResult } from "shared_types/types/shared_types";
import {
  HttpResponse,
  HttpHeader,
  HttpResultVariantOk,
} from "shared_types/types/shared_types";

export async function request({
  url,
  method,
  headers,
}: HttpRequest): Promise<HttpResult> {
  const request = new Request(url, {
    method,
    headers: headers.map((header) => [header.name, header.value]),
  });

  const response = await fetch(request);

  const responseHeaders = Array.from(
    response.headers.entries(),
    ([name, value]) => new HttpHeader(name, value)
  );

  const body = await response.arrayBuffer();

  return new HttpResultVariantOk(
    new HttpResponse(
      response.status,
      responseHeaders,
      Array.from(new Uint8Array(body))
    )
  );
}
